### ISSUE TITLE
Data race in middleware/closenotify

### ISSUE THREAD

**Issue by diegobernardes** [pressly/chi#141]
Using the close notify middleware is giving me data races. I'm hitting ctrl+r like crazy in the browser.

This is the only part of the handler that writes in http.ResponseWriter, in case of context done, the program don't do anything. 

```
select {
case <-r.Context().Done():
  fmt.Println("end of request")
case err := <-chanErr:
  c.sendInvalidRequest(w, err)
case resp := <-chanResp:
  c.sendResponse(r, w, resp)
}

return
```

But the last request get processed because the browser tab is open, then the headers are writted and this error popup on console:

```2017/01/21 10:18:32 http: multiple response.WriteHeader calls```

Am i doing anything wrong? 

#### System details

```
go version go1.8rc2 linux/amd64
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/diego/projects/go/1.8rc2"
GORACE=""
GOROOT="/home/diego/programs/go/1.8rc2"
GOTOOLDIR="/home/diego/programs/go/1.8rc2/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build796800410=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
GOROOT/bin/go version: go version go1.8rc2 linux/amd64
GOROOT/bin/go tool compile -V: compile version go1.8rc2 X:framepointer
uname -sr: Linux 4.8.13-1-ARCH
/usr/lib/libc.so.6: GNU C Library (GNU libc) stable release version 2.24, by Roland McGrath et al.
```

**Comment by diegobernardes** [comment]
Forgot the data race message:

```
==================
WARNING: DATA RACE
Read at 0x00c4201a01e8 by goroutine 25:
  net/http.(*response).finishRequest()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1534 +0x57
  net/http.(*conn).serve()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1830 +0x76b

Previous write at 0x00c4201a01e8 by goroutine 32:
  net/http.(*response).WriteHeader()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1056 +0xb9
  stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware.CloseNotify.func1.1()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware/closenotify.go:25 +0x6f

Goroutine 25 (running) created at:
  net/http.(*Server).Serve()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2668 +0x35f
  net/http.(*Server).ListenAndServe()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2585 +0xe0
  net/http.ListenAndServe()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2787 +0xee
  stash.b2w/alpha/omega.server()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/server.go:96 +0x4bb
  stash.b2w/alpha/omega.Server()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/omega.go:11 +0xca
  main.main.func1()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/cli/omega.go:27 +0xc9
  stash.b2w/alpha/omega/vendor/github.com/spf13/cobra.(*Command).execute()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/spf13/cobra/command.go:636 +0x85d
  stash.b2w/alpha/omega/vendor/github.com/spf13/cobra.(*Command).ExecuteC()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/spf13/cobra/command.go:722 +0x46b
  stash.b2w/alpha/omega/vendor/github.com/spf13/cobra.(*Command).Execute()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/spf13/cobra/command.go:681 +0x38
  main.main()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/cli/omega.go:44 +0x794

Goroutine 32 (finished) created at:
  stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware.CloseNotify.func1()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware/closenotify.go:28 +0xc5
  net/http.HandlerFunc.ServeHTTP()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1942 +0x51
  stash.b2w/alpha/omega/vendor/github.com/pressly/chi.(*Mux).ServeHTTP()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/pressly/chi/mux.go:79 +0x37f
  net/http.serverHandler.ServeHTTP()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2568 +0xbc
  net/http.(*conn).serve()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1825 +0x716
==================
==================
WARNING: DATA RACE
Read at 0x00c4201a0230 by goroutine 25:
  net/http.(*chunkWriter).writeHeader()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1200 +0x21c7
  net/http.(*chunkWriter).close()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:379 +0x17a
  net/http.(*response).finishRequest()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1540 +0xc1
  net/http.(*conn).serve()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1830 +0x76b

Previous write at 0x00c4201a0230 by goroutine 32:
  net/http.(*response).WriteHeader()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1057 +0xd2
  stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware.CloseNotify.func1.1()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware/closenotify.go:25 +0x6f

Goroutine 25 (running) created at:
  net/http.(*Server).Serve()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2668 +0x35f
  net/http.(*Server).ListenAndServe()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2585 +0xe0
  net/http.ListenAndServe()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2787 +0xee
  stash.b2w/alpha/omega.server()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/server.go:96 +0x4bb
  stash.b2w/alpha/omega.Server()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/omega.go:11 +0xca
  main.main.func1()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/cli/omega.go:27 +0xc9
  stash.b2w/alpha/omega/vendor/github.com/spf13/cobra.(*Command).execute()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/spf13/cobra/command.go:636 +0x85d
  stash.b2w/alpha/omega/vendor/github.com/spf13/cobra.(*Command).ExecuteC()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/spf13/cobra/command.go:722 +0x46b
  stash.b2w/alpha/omega/vendor/github.com/spf13/cobra.(*Command).Execute()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/spf13/cobra/command.go:681 +0x38
  main.main()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/cli/omega.go:44 +0x794

Goroutine 32 (finished) created at:
  stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware.CloseNotify.func1()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/pressly/chi/middleware/closenotify.go:28 +0xc5
  net/http.HandlerFunc.ServeHTTP()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1942 +0x51
  stash.b2w/alpha/omega/vendor/github.com/pressly/chi.(*Mux).ServeHTTP()
      /home/diego/projects/go/1.8rc2/src/stash.b2w/alpha/omega/vendor/github.com/pressly/chi/mux.go:79 +0x37f
  net/http.serverHandler.ServeHTTP()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:2568 +0xbc
  net/http.(*conn).serve()
      /home/diego/programs/go/1.8rc2/src/net/http/server.go:1825 +0x716
==================
```

**Comment by pkieltyka** [comment]
@diegobernardes hey there - can you provide a small gist example that shows the problem? Im not sure what chanErr and chanResp are in the first example. I'll give it a review though

**Comment by pkieltyka** [comment]
I tried this little program

```go
func main() {
	r := chi.NewRouter()
	r.Use(middleware.RequestID)
	r.Use(middleware.Logger)
	r.Use(middleware.CloseNotify)
	r.Get("/", func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(1 * time.Second)
		w.Write([]byte("hello world"))
	})
	http.ListenAndServe(":3333", r)
}
```

and likewise in the browser reloaded the page before the request could finish (simulating a disconnect), and I couldnt get a race to occur

**Comment by diegobernardes** [comment]
Yes, your examples does not give any error, but try to comment the logger and the sleep:

```go
package main

import (
	"net/http"

	"github.com/pressly/chi"
	"github.com/pressly/chi/middleware"
)

func main() {
	r := chi.NewRouter()
	r.Use(middleware.RequestID)
	// r.Use(middleware.Logger)
	r.Use(middleware.CloseNotify)
	r.Get("/", func(w http.ResponseWriter, r *http.Request) {
    // time.Sleep(1 * time.Second)
		w.Write([]byte("hello world"))
	})
	http.ListenAndServe(":3333", r)
}
```

Using wrk:
```
wrk -c 100 -d 10s http://localhost:3333
```

Give this results:
```
go run main.go
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
2017/01/23 09:36:21 http: multiple response.WriteHeader calls
```

About the race condition, don't know why, but i can not reproduce...
Maybe was the middlewares or the order of then that caused that error.

**Comment by pkieltyka** [comment]
Thanks @diegobernardes I'll fix it soon

**Comment by pkieltyka** [comment]
I believe the fix is very easy, which is to just comment out https://github.com/pressly/chi/blob/master/middleware/closenotify.go#L25 because its really just for internal communication to the logger to print disconnected, its not that useful. I will update the behaviour. From that point, I do believe middleware.CloseNotify() will effectively be a no-op in go 1.8, perhaps panic that it should be be used 1.8+ because the default behaviour is to cancel a context when a client disconnects.

**Comment by diegobernardes** [comment]
Didn't know that in 1.8 the CloseNotify cancel the context, good to know.
Thanks!

**Comment by pkieltyka** [comment]
Resolved by 7765b0916fdc6791938554e9d53d39cae059ac8c
